### PR TITLE
Fix pos_label not passed to ROC/PR curve artifacts in sklearn autolog

### LINKED ISSUE #6797: [BUG] pos_label param in mlflow.sklearn.autolog not working properly
### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

1.28.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Monterrey 12.5.1 (chip Apple M1 Pro)
- **Python version**: 3.8.13
- **yarn version, if running the dev UI**: 

### Folder structure

```
./
📄 conda.yml
📁 bug_report/
├── 📄 MLproject
├── 📄 conda.yml
├── 📄 run.py
└── 📄 sample.csv
```

### command to execute the code

`mlflow run -e bug_report/run.py --experiment-name bug_report  .`

### main `conda.yml` file to execute run command:

```yaml
name: env_main
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.10.6
  - pip=22.2.2
  - hydra-core=1.2.0
  - dask=2022.8.1
  - black=22.6.0
  - flake8=5.0.4
  - jupyter=1.0.0
  - scikit-learn=1.1.2
  - seaborn=0.12.0
  - pip:
      - mlflow==1.28.0
      - dvc==2.23.0
      - dvc-s3==2.19.0

```

### bug_report `bug_report/conda.yml` file:

```yaml
name: env_bug_report
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.8.13
  - pip=22.2.2
  - pandas=1.2.5
  - dask=2021.10.0
  - typer=0.6.1
  - scikit-learn=1.1.2
  - matplotlib=3.5.3
  - pip:
      - mlflow==1.28.0

```

### MLproject

```MLproject
name: bug_report
conda_env: conda.yml

entry_points:
  main:
    parameters:
      path-train:
        type: str

    command: >-
      python run.py --path-train {path-train}

```

```python
import logging

import mlflow.sklearn
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd

logging.basicConfig(level=logging.INFO, format="%(asctime)-15s %(message)s")
logger = logging.getLogger()

##### BAD EXAMPLE #####
# - contains two distinct values in y_train: `Charged off` and `Fully Paid`
# - autolog has parameter `pos_label="Charged Off"`
# - detects pos_label as `Fully Paid`.
logger.info("Bad example")
logger.info("Training model with error in pos_label")

df_train = pd.read_csv("bug_report/sample.csv")
X_train = df_train.drop(columns=["loan_status"])
y_train = np.array(df_train["loan_status"])
logger.info(f"...unique values in y_train {np.unique(y_train)}")
model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Charged Off")
with mlflow.start_run() as _:
    model.fit(X_train, y_train)

##### GOOD EXAMPLE #####
# - contains two distinct values in y_train: `Negative_label` and `Positive_label`
# - autolog has parameter `pos_label="Positive_label"`
# - detects pos_label as `Positive_label`.
logger.info("Good example")
logger.info("Training model without error in pos_label")
X_train_2 = np.array([[1, 0], [0, 1], [1, 0], [1, 1], [1, 1], [1, 2]])
y_train_2 = np.array(
    [
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Positive_label",
        "Positive_label",
    ]
)
logger.info(f"...unique values in y_train {np.unique(y_train_2)}")

model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Positive_label")
with mlflow.start_run() as _:
    model.fit(X_train_2, y_train_2)

```

### `sample.csv` file

I let the data in Google Drive to repeat the mistake.

https://drive.google.com/file/d/1kH-i1_y2B-2vV0KJhpT52vhDs7IpZA3B/view?usp=sharing

### Describe the problem

I use the line `mlflow.sklearn.autolog(pos_label="Charged Off")` to define the positive label (this is the minority class), but the positive label reported in the mlflow server is `"Fully Paid"`.

Evidence:
image

I created dummy samples to detect the problem, but the error didn't happen again. I let the data in Google Drive to repeat the mistake.

https://drive.google.com/file/d/1kH-i1_y2B-2vV0KJhpT52vhDs7IpZA3B/view?usp=sharing

### Tracking information

_No response_

### Code to reproduce issue

```python
import logging

import mlflow.sklearn
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd

logging.basicConfig(level=logging.INFO, format="%(asctime)-15s %(message)s")
logger = logging.getLogger()

##### BAD EXAMPLE #####
logger.info("Bad example")
logger.info("Training model with error in pos_label")
# - contains two distinct values in y_train: `Charged off` and `Fully Paid`
# - autolog has parameter `pos_label="Charged Off"`
# - detects pos_label as `Fully Paid`.
df_train = pd.read_csv("bug_report/sample.csv")
X_train = df_train.drop(columns=["loan_status"])
y_train = np.array(df_train["loan_status"])
logger.info(f"...unique values in y_train {np.unique(y_train)}")
model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Charged Off")
with mlflow.start_run() as _:
    model.fit(X_train, y_train)

##### GOOD EXAMPLE #####
logger.info("Good example")
logger.info("Training model without error in pos_label")
X_train_2 = np.array([[1, 0], [0, 1], [1, 0], [1, 1], [1, 1], [1, 2]])
y_train_2 = np.array(
    [
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Positive_label",
        "Positive_label",
    ]
)
logger.info(f"...unique values in y_train {np.unique(y_train_2)}")

model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Positive_label")
with mlflow.start_run() as _:
    model.fit(X_train_2, y_train_2)

```

### Stack trace

```python
# by executing this command mlflow run -e bug_report/run.py --experiment-name bug_repor
```

```python
2022/09/14 14:07:33 INFO mlflow.utils.conda: Conda environment mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77 already exists.
2022/09/14 14:07:33 INFO mlflow.projects.utils: === Created directory /var/folders/nt/w74p_l0x1h1g63lfz0ybhrdr0000gp/T/tmptrl1l7j8 for downloading remote URIs passed to arguments of type 'path' ===
2022/09/14 14:07:33 INFO mlflow.projects.backend.local: === Running command 'source /Users/nelsoncardenas/opt/miniconda3/bin/../etc/profile.d/conda.sh && conda activate mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77 1>&2 && python bug_report/run.py' in run with ID 'ec5922588def43dea40f502f5482245a' === 
2022-09-14 14:07:37,424 Bad example
2022-09-14 14:07:37,424 Training model with error in pos_label
2022-09-14 14:07:37,457 ...unique values in y_train ['Charged Off' 'Fully Paid']
2022/09/14 14:07:44 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/nelsoncardenas/opt/miniconda3/envs/mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77/lib/python3.10/site-packages/mlflow/models/signature.py:129: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2022/09/14 14:07:45 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/nelsoncardenas/opt/miniconda3/envs/mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils."
2022-09-14 14:07:45,745 Good example
2022-09-14 14:07:45,745 Training model without error in pos_label
2022-09-14 14:07:45,745 ...unique values in y_train ['Negative_label' 'Positive_label']
2022/09/14 14:07:47 INFO mlflow.projects: === Run (ID 'ec5922588def43dea40f502f5482245a') succeeded ===
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by vedjaw**
### Related Issues/PRs

Closes #6797

### What changes are proposed in this pull request?

`mlflow.sklearn.autolog(pos_label=...)` correctly used `pos_label` for scalar
classification metrics, but autologged ROC and precision-recall curve artifacts
did not receive `pos_label`. For binary classifiers whose positive class is not
`classes_[1]`, the logged curves were computed for the wrong class.

This threads `pos_label` through `_get_classifier_artifacts()` into the sklearn
ROC / precision-recall plot calls.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

Added:
- `test_get_classifier_artifacts_respects_pos_label`
- `test_autolog_pos_label_used_for_classifier_artifact_curves`

```bash
pytest tests/sklearn/test_sklearn_autolog.py::test_get_classifier_artifacts_respects_pos_label
pytest tests/sklearn/test_sklearn_autolog.py::test_autolog_pos_label_used_for_classifier_artifact_curves
```

### Does this PR require documentation update?

- [x] No.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.
- [ ] Yes. Please link the corresponding PR or explain how you plan to update it.

#### Is this a user-facing change?

- [ ] No.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixed a bug where `mlflow.sklearn.autolog(pos_label=...)` did not pass
`pos_label` to autologged ROC and precision-recall curve artifacts, causing
incorrect curves for binary classifiers when the positive class was not
`classes_[1]`.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

File: tests/sklearn/test_sklearn_autolog.py [hunk 1]
```diff
@@ -1838,6 +1839,61 @@ def test_autolog_pos_label_used_for_training_metric():
...
+            "Positive_label",
+            "Positive_label",
+        ]
+    )
+    model = sklearn.linear_model.LogisticRegression()
+    model.fit(X, y)
+
+    artifacts = _get_classifier_artifacts(model, "training_", X, y, None, "Positive_label")
+    roc_artifact = next(a for a in artifacts if a.name == "training_roc_curve")
+    pr_artifact = next(a for a in artifacts if a.name == "training_precision_recall_curve")
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply tests/sklearn/test_sklearn_autolog.py:1861]: Fixed in b4a2671 — added a `skipif` on `_is_plotting_supported()`, since the early return makes both `next(...)` lookups raise `StopIteration` rather than skip. The curve calls in the body now run unconditionally.

File: mlflow/sklearn/utils.py [hunk 2]
```diff
@@ -304,7 +304,7 @@ def _get_class_labels_from_estimator(estimator):
     return estimator.classes_ if hasattr(estimator, "classes_") else None
 
 
-def _get_classifier_artifacts(fitted_estimator, prefix, X, y_true, sample_weight):
+def _get_classifier_artifacts(fitted_estimator, prefix, X, y_true, sample_weight, pos_label):
     """
     Draw and record various common artifacts for classifier
```
[the remark this answers is not in the document]
Comment (vedjaw) [reply mlflow/sklearn/utils.py:310]: Fixed in b4a2671 — docstring now documents `prefix`, `X`, `y_true`, `sample_weight` and `pos_label`, notes that `pos_label` applies to the precision recall and roc curves but not the confusion matrix, and that the return is empty when plotting is unsupported. Dropped the stale `fit_args`/`fit_kwargs` and "regressor" wording.
